File: packages/websearch/src/engines/html.ts
/**
 * Minimal, dependency-free HTML text utilities for the scrape-based engines
 * (Mojeek) and tagged-snippet APIs (Wikipedia's `<span class=searchmatch>`).
 *
 * We deliberately do NOT pull in jsdom/readability here (unlike webfetch):
 * websearch parses a small, known result-list structure, not arbitrary
 * article bodies, so a targeted entity-decode + tag-strip keeps the package
 * light and the parse fast. The Mojeek block parser lives in its engine.
 */

const NAMED_ENTITIES: Readonly<Record<string, string>> = {
  amp: "&",
  lt: "<",
  gt: ">",
  quot: '"',
  apos: "'",
  nbsp: " ",
  rsaquo: "\u203a",
  lsaquo: "\u2039",
  raquo: "\u00bb",
  laquo: "\u00ab",
  hellip: "\u2026",
  mdash: "\u2014",
  ndash: "\u2013",
  rsquo: "\u2019",
  lsquo: "\u2018",
  ldquo: "\u201c",
  rdquo: "\u201d",
  middot: "\u00b7",
  deg: "\u00b0",
  copy: "\u00a9",
  reg: "\u00ae",
  trade: "\u2122",
  eacute: "\u00e9",
  egrave: "\u00e8",
  agrave: "\u00e0",
  ccedil: "\u00e7",
  uuml: "\u00fc",
  ouml: "\u00f6",
  auml: "\u00e4",
};

/** Decode the HTML entities that actually occur in SERP markup. */
export function decodeEntities(input: string): string {
  return input.replace(/&(#x?[0-9a-fA-F]+|[a-zA-Z][a-zA-Z0-9]*);/g, (m, body) => {
    const b = body as string;
    if (b.charAt(0) === "#") {
      const isHex = b.charAt(1) === "x" || b.charAt(1) === "X";
      const code = Number.parseInt(b.slice(isHex ? 2 : 1), isHex ? 16 : 10);
      if (Number.isFinite(code) && code > 0 && code <= 0x10ffff) {
        try {
          return String.fromCodePoint(code);
        } catch {
          return m;
        }
      }
      return m;
    }
    const named = NAMED_ENTITIES[b.toLowerCase()];
    return named ?? m;
  });
}

/** Strip HTML tags, decode entities, and collapse whitespace. */
export function stripTags(html: string): string {
  const noTags = html.replace(/<[^>]*>/g, " ");
  return decodeEntities(noTags).replace(/\s+/g, " ").trim();
}

File: crates/websearch/src/engines/html.rs
//! Minimal, dependency-free HTML text utilities for the scrape-based engine
//! (Mojeek) and tagged-snippet APIs (Wikipedia). Mirrors TS `engines/html.ts`:
//! decode the entities that occur in SERP markup, strip tags, collapse
//! whitespace. No scraper/html5ever dependency — the parse target is a small,
//! known result-list structure, not arbitrary article bodies.

/// Decode the HTML entities that actually occur in SERP markup.
pub(crate) fn decode_entities(input: &str) -> String {
    let mut out = String::with_capacity(input.len());
    let bytes = input.as_bytes();
    let mut i = 0;
    while i < bytes.len() {
        if bytes[i] == b'&' {
            if let Some(semi) = input[i + 1..].find(';') {
                let body = &input[i + 1..i + 1 + semi];
                if let Some(decoded) = decode_one(body) {
                    out.push_str(&decoded);
                    i += semi + 2; // skip '&' .. ';'
                    continue;
                }
            }
        }
        // Push this UTF-8 char whole.
        let ch = input[i..].chars().next().unwrap();
        out.push(ch);
        i += ch.len_utf8();
    }
    out
}

fn decode_one(body: &str) -> Option<String> {
    if let Some(rest) = body.strip_prefix('#') {
        let code = if let Some(hex) = rest.strip_prefix('x').or_else(|| rest.strip_prefix('X')) {
            u32::from_str_radix(hex, 16).ok()?
        } else {
            rest.parse::<u32>().ok()?
        };
        return char::from_u32(code).map(|c| c.to_string());
    }
    let s = match body.to_ascii_lowercase().as_str() {
        "amp" => "&",
        "lt" => "<",
        "gt" => ">",
        "quot" => "\"",
        "apos" => "'",
        "nbsp" => " ",
        "rsaquo" => "\u{203a}",
        "lsaquo" => "\u{2039}",
        "raquo" => "\u{00bb}",
        "laquo" => "\u{00ab}",
        "hellip" => "\u{2026}",
        "mdash" => "\u{2014}",
        "ndash" => "\u{2013}",
        "rsquo" => "\u{2019}",
        "lsquo" => "\u{2018}",
        "ldquo" => "\u{201c}",
        "rdquo" => "\u{201d}",
        "middot" => "\u{00b7}",
        "deg" => "\u{00b0}",
        "copy" => "\u{00a9}",
        "reg" => "\u{00ae}",
        "trade" => "\u{2122}",
        "eacute" => "\u{00e9}",
        "egrave" => "\u{00e8}",
        "agrave" => "\u{00e0}",
        "ccedil" => "\u{00e7}",
        "uuml" => "\u{00fc}",
        "ouml" => "\u{00f6}",
        "auml" => "\u{00e4}",
        _ => return None,
    };
    Some(s.to_string())
}

/// Strip HTML tags, decode entities, and collapse whitespace.
pub(crate) fn strip_tags(html: &str) -> String {
    let mut no_tags = String::with_capacity(html.len());
    let mut in_tag = false;
    for ch in html.chars() {
        match ch {
            '<' => in_tag = true,
            '>' => {
                in_tag = false;
                no_tags.push(' ');
            }
            _ if !in_tag => no_tags.push(ch),
            _ => {}
        }
    }
    let decoded = decode_entities(&no_tags);
    decoded.split_whitespace().collect::<Vec<_>>().join(" ")
}
